magma/passes/unused_undriven.py

```diff
@@ -0,0 +1,32 @@
+from abc import abstractmethod
+from .passes import CircuitPass
+
+
+class UndrivenUnusedPass(CircuitPass):
+    @abstractmethod
+    def process(self, port_values):
+        raise NotImplementedError()
+
+    def __call__(self, circuit):
+        with circuit.open():
+            self.process(circuit, circuit.interface.ports.values())
+            for instance in circuit.instances:
+                self.process(circuit, instance.interface.ports.values())
+
+
+class DriveUndriven(UndrivenUnusedPass):
+    def process(self, circuit, port_values):
+        for value in port_values:
+            if value.is_input() and value.trace() is None:
+                # Turn decls into definition
+                circuit._is_definition = True
```

Comment:
as an aside, this is why we need to revisit the circuit machinery when editing becomes possible. There's potentially a lot of things which could get out of sync, and `_is_definition` is one of them.